resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -229,8 +229,17 @@ private[spark] class ApplicationMaster(
         }
 
         if (!unregistered) {
+          // Only when ApplicationMaster's exit code is "EXIT_REPORTER_FAILURE" or "EXIT_EARLY" or
+          // "EXIT_UNCAUGHT_EXCEPTION" we should let RM to do reattempt. If the exception is from
+          // Spark internally, for example illegal state, executor failure, user code bug, we
+          // should not trigger another attempt.
+          val shouldUnregister = !(exitCode == ApplicationMaster.EXIT_REPORTER_FAILURE ||
+            exitCode == ApplicationMaster.EXIT_EARLY ||
+            exitCode == ApplicationMaster.EXIT_UNCAUGHT_EXCEPTION)
```

Comment:
The `exitCode` here is only from pyspark or R runner right ?
Have we determined that there is no overlap in `exitCode` ?

I ask because IIRC spark was not using the `exitCode` until now (other than to communicate it to yarn) - but we will now start assigning meaning to it, which can overlap/conflict with pyspark or R.

If there is no overlap, it is fine to do this (with the negation as mentioned).
If there is potential for overlap, we should do this only when exitCode is not coming from SparkUserAppException ?